[docs] Handbook pages (#1141)
mui/base-ui: docs/src/app/(public)/(content)/react/handbook/styling/page.mdx

# Styling

<Subtitle>A guide to styling Base UI components with your preferred styling engine.</Subtitle>
<Meta
  name="description"
  content="Learn how to make the most of Base UI's accessibility features and guidelines."
/>

Base UI components are unstyled, don't bundle CSS, and are compatible with Tailwind, CSS Modules, CSS-in-JS, or any other styling solution you prefer. You retain total control of your styling layer.

## Style hooks

> ### CSS Classes
> 
> Every component accepts a `className` prop which you can use to pass Tailwind classes or custom CSS classes to its DOM element. The `className` prop, in addition to accepting a string, can be defined as a function that takes a component state as an argument:

Reviewer: Formatting nit/suggestion

```suggestion
Every component accepts a `className` prop which you can use to pass CSS classes to its DOM element.

The `className` prop, in addition to accepting a string, can be defined as a function that takes the component state as an argument:
```

```tsx title="switch.tsx"
<Switch.Thumb className={(state) => (state.checked ? 'checked' : 'unchecked')} />
```

> ### Data attributes
> 
> Data attributes are provided as style hooks. Component states can be styled with data attributes such as `[data-open]`. Certain props can be targeted with data attributes such as `[data-side]`. Some components provide `[data-starting-style]` and `[data-ending-style]` attributes for styling CSS transitions. Check out each component's API reference for a complete list of available data attributes.

Reviewer: A bit much, I'd simplify

```suggestion
 ### Data attributes

Base UI components provide data attributes designed for styling their states. For example, a [Switch](/react/components/switch) may be styled using its `[data-checked]` and `[data-unchecked]` attributes, among others.

Check out each component's API reference for a complete list of available data attributes.
```

> ### CSS custom properties
> 
> Some components expose CSS custom properties to aid with styling, such as `--accordion-panel-height`, `--transform-origin`, and `--active-tab-width`.

Reviewer: - Totally a nit, but would feel better in the side nav if we put CSS classes and CSS vars closer together
- I'd call this section "CSS variables" since it's the colloquial name that most people would rather recognise

```suggestion
### CSS custom properties

Some components expose CSS custom properties to aid with styling, such as `--accordion-panel-height`, `--transform-origin`, or `--active-tab-width`.
```

## Tailwind

Apply Tailwind classes to each part via the `className` prop.

> ```tsx title="index.tsx"

Reviewer: ```suggestion
```tsx title="menu.tsx"
```

            <Menu.Item className="data-[highlighted]:before:-z-1 flex cursor-default select-none py-2 pl-4 pr-8 text-sm leading-4 outline-0 data-[highlighted]:relative data-[highlighted]:z-0 data-[highlighted]:text-gray-50 data-[highlighted]:before:absolute data-[highlighted]:before:inset-x-1 data-[highlighted]:before:inset-y-0 data-[highlighted]:before:rounded-sm data-[highlighted]:before:bg-gray-900">
              Add to Library
            </Menu.Item>
            <Menu.Item className="data-[highlighted]:before:-z-1 flex cursor-default select-none py-2 pl-4 pr-8 text-sm leading-4 outline-0 data-[highlighted]:relative data-[highlighted]:z-0 data-[highlighted]:text-gray-50 data-[highlighted]:before:absolute data-[highlighted]:before:inset-x-1 data-[highlighted]:before:inset-y-0 data-[highlighted]:before:rounded-sm data-[highlighted]:before:bg-gray-900">
              Add to Playlist
            </Menu.Item>
          </Menu.Popup>
        </Menu.Positioner>
      </Menu.Portal>
    </Menu.Root>
  );
}
```

## CSS Modules

Apply custom CSS classes to each part via the `className` prop. Then style those classes in a CSS Modules file.

> ```tsx title="index.tsx"

Reviewer: ```suggestion
```tsx title="menu.tsx"
```
